lib/chef_handler_foreman/foreman_facts.rb

```diff
@@ -85,11 +97,27 @@ def get_key(key, prefix)
     end
 
     def send_attributes(attributes)
-      if uploader
-        uploader.foreman_request('/api/hosts/facts', attributes, node.name)
+      if @cache_file and !@cache_expired
+        Chef::Log.info "No attributes have changed - not uploading to foreman" 
       else
-        Chef::Log.error "No uploader registered for foreman facts, skipping facts upload"
+        if uploader
+          uploader.foreman_request('/api/hosts/facts', attributes, node.name)
+        else
+          Chef::Log.error "No uploader registered for foreman facts, skipping facts upload"
+        end
+      end
+    end
+
+    def verify_checksum(attributes)
+      @cache_expired = true
+      attrs_checksum = Digest::MD5.hexdigest(attributes.to_s)
+      if File.exist?(@cache_file)
+        contents = File.read(@cache_file)
+        if attrs_checksum == contents
+          @cache_expired = false
+        end
       end
+      File.open(@cache_file, 'w') { |f| f.write(attrs_checksum) }
```

Comment:
during testing I realized that if user set a bad path (e.g. /va/cache) we get an exception on this line, I think we should just rescue here, we can safely ignore errors, in worst case facts won't be cached

```ruby
def verify_checksum(attributes)
  $code...
rescue => e
  @cache_expired = true
  Chef::Log.info "unable to verify cache checksum - #{e.message}, facts will be sent"
  Chef::Log.debug e.backtrace.join("\n")
end
```